File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AvatarInterestsReply.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * AvatarInterestsReply
 *
 * <p>Template: {@code AvatarInterestsReply Low 172 Trusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code LLAvatarPropertiesProcessor::processAvatarInterestsReply()} in indra/newview/llavatarpropertiesprocessor.cpp
 * (secondlife/viewer @ c179f76c01).
 */
public class AvatarInterestsReply extends SLMessage {
    public AgentData AgentData_Field;
    public PropertiesData PropertiesData_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID - your id
        public UUID AvatarID; // LLUUID - avatar you're asking about
    }

    /** Block PropertiesData, Single. */
    public static class PropertiesData {
        public byte[] LanguagesText; // Variable 1 - string
        public int SkillsMask; // U32
        public byte[] SkillsText; // Variable 1 - string
        public int WantToMask; // U32
        public byte[] WantToText; // Variable 1 - string
    }

    public AvatarInterestsReply() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
        this.PropertiesData_Field = new PropertiesData();
    }

    @Override
    public int CalcPayloadSize() {
        return this.PropertiesData_Field.WantToText.length + 5 + 4 + 1 + this.PropertiesData_Field.SkillsText.length + 1 + this.PropertiesData_Field.LanguagesText.length + 36;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleAvatarInterestsReply(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 172 (AvatarInterestsReply).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0xAC);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.AvatarID);
        packInt(byteBuffer, this.PropertiesData_Field.WantToMask);
        packVariable(byteBuffer, this.PropertiesData_Field.WantToText, 1);
        packInt(byteBuffer, this.PropertiesData_Field.SkillsMask);
        packVariable(byteBuffer, this.PropertiesData_Field.SkillsText, 1);
        packVariable(byteBuffer, this.PropertiesData_Field.LanguagesText, 1);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.AvatarID = unpackUUID(byteBuffer);
        this.PropertiesData_Field.WantToMask = unpackInt(byteBuffer);
        this.PropertiesData_Field.WantToText = unpackVariable(byteBuffer, 1);
        this.PropertiesData_Field.SkillsMask = unpackInt(byteBuffer);
        this.PropertiesData_Field.SkillsText = unpackVariable(byteBuffer, 1);
        this.PropertiesData_Field.LanguagesText = unpackVariable(byteBuffer, 1);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AvatarInterestsReply.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * AvatarInterestsReply
 *
 * <p>Template: {@code AvatarInterestsReply Low 172 Trusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code LLAvatarPropertiesProcessor::processAvatarInterestsReply()} in indra/newview/llavatarpropertiesprocessor.cpp
 * (secondlife/viewer @ c179f76c01).
 */
open class AvatarInterestsReply : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var PropertiesData_Field: PropertiesData = PropertiesData()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var AvatarID: UUID? = null
    }

    /** Block PropertiesData, Single. */
    open class PropertiesData {
        @JvmField var LanguagesText: ByteArray? = null
        @JvmField var SkillsMask: Int = 0
        @JvmField var SkillsText: ByteArray? = null
        @JvmField var WantToMask: Int = 0
        @JvmField var WantToText: ByteArray? = null
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return PropertiesData_Field.WantToText!!.size + 5 + 4 + 1 + PropertiesData_Field.SkillsText!!.size + 1 + PropertiesData_Field.LanguagesText!!.size + 36
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleAvatarInterestsReply(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 172 (AvatarInterestsReply).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0xAC).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.AvatarID)
        packInt(byteBuffer, PropertiesData_Field.WantToMask)
        packVariable(byteBuffer, PropertiesData_Field.WantToText, 1)
        packInt(byteBuffer, PropertiesData_Field.SkillsMask)
        packVariable(byteBuffer, PropertiesData_Field.SkillsText, 1)
        packVariable(byteBuffer, PropertiesData_Field.LanguagesText, 1)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.AvatarID = unpackUUID(byteBuffer)
        PropertiesData_Field.WantToMask = unpackInt(byteBuffer)
        PropertiesData_Field.WantToText = unpackVariable(byteBuffer, 1)
        PropertiesData_Field.SkillsMask = unpackInt(byteBuffer)
        PropertiesData_Field.SkillsText = unpackVariable(byteBuffer, 1)
        PropertiesData_Field.LanguagesText = unpackVariable(byteBuffer, 1)
    }
}
